include/gridtools/fn/extents.hpp

```diff
@@ -0,0 +1,107 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+
+#include "../common/int_vector.hpp"
+#include "../common/integral_constant.hpp"
+#include "../common/utility.hpp"
+#include "../meta.hpp"
+
+namespace gridtools {
+    namespace fn {
+        template <class Dim, ptrdiff_t L, ptrdiff_t U>
+        struct extent {
+            static_assert(L <= U);
+            using dim_t = Dim;
+            using lower_t = integral_constant<ptrdiff_t, L>;
+            using upper_t = integral_constant<ptrdiff_t, U>;
+            using size_t = integral_constant<std::size_t, U - L>;
+            using list_t = meta::list<Dim, lower_t, upper_t>;
+        };
+
+        template <class>
+        struct is_extent : std::false_type {};
+
+        template <class Dim, ptrdiff_t L, ptrdiff_t U>
+        struct is_extent<extent<Dim, L, U>> : std::true_type {};
+
+#ifdef __cpp_concepts
+        template <class T>
+        concept extent_c = is_extent<T>::value;
+#endif
+
+        template <class... Ts>
+        struct extents {
+            static_assert(meta::all_of<is_extent, meta::list<Ts...>>::value);
+            static_assert(meta::is_set<meta::list<typename Ts::dim_t...>>::value);
+            using keys_t = hymap::keys<typename Ts::dim_t...>;
+            static_assert(is_int_vector<typename keys_t::template values<typename Ts::lower_t...>>::value);
+            static_assert(is_int_vector<typename keys_t::template values<typename Ts::upper_t...>>::value);
+            static_assert(is_int_vector<typename keys_t::template values<typename Ts::size_t...>>::value);
+
+            static GT_CONSTEVAL auto offsets() {
+                return int_vector::prune_zeros(typename keys_t::template values<typename Ts::lower_t...>());
+            }
+            static GT_CONSTEVAL auto sizes() {
+                return int_vector::prune_zeros(typename keys_t::template values<typename Ts::size_t...>());
+            }
+        };
+
+        template <class T>
+        using is_extents = meta::is_instantiation_of<extents, T>;
```

Comment:
add `meta::all_of<is_extent, meta::list<Ts...>>::value` and `meta::is_set<meta::list<typename Ts::dim_t...>>::value`